### PR TITLE
Bulk Linking of Top Containers to Archival Objects

### PR DESCRIPTION
**Pull request by ktamaral**
Note: This is a **replacement** PR for #1944. The exiting PR #1944 can be closed.

The original feature branch "top_container_upload_download" is in the existing PR #1944 and that branch is now outdated. This pull request contains the new branch "top_container_upload_download_rebase_fixes".  This new branch was rebased on the current version of archivesspace/master. We decided to open a new PR for the newly rebased branch because otherwise we would've had to go through the entire rebase process again on the original feature branch. In summary, this new branch is rebased on master and contains all of the requested fixes from the last review.

## Description

Allows users to link top containers to archival objects in bulk at the resource level using CSV or XLSX. Linking can be by either an existing top container ID or by creating a new top container in the spreadsheet. To do this users:

1. Export a Top Container Template form
2. Fill out the fields of the csv form.
Some fields are prepopulated and should not be touched
Minimum required user entered fields are:
a. Top Container Record Number (internal id) XOR (Top Container Indicator AND Top Container Type)
b. Instance Type
NOTE: Location ID and Container Profile ID can only link to new top containers. Existing top containers will not be updated.
3. Use the "Link to Spreadsheet" button on the resource edit screen, use the file selector to select the proper file, and enable the toggle button "Link Top Containers".
To link the containers, click "Import from Spreadsheet".
Enable the "only validate" checkbox to validate only without linking.
This change was necessary because it allows users to link top containers in bulk instead of performing the mundane task of doing this one by one.

## Related JIRA Ticket or GitHub Issue

The requested fixes from the original review are tracked in this google spreadsheet. Notes have been added to explain what was done for each issue.
https://docs.google.com/spreadsheets/d/1zP4_Cz6EzITo9q8DXe_VbuzDFwrCnC_RlUUSHEI7zUI/edit#gid=0

## How Has This Been Tested?

RSpec tests to test validation and uploading
Manual front end testing by developers and end users at Harvard.
## Screenshots (if appropriate):
Download template link:
tc_upload_download_1

Link Top Containers toggle button has been added to the existing bulk import modal. This modal appears when clicking on the "Link via Spreadsheet" when in edit mode at the resource level.
tc_upload_download_2

When the background job is complete, the log displays any errors that may have occurred, and if no errors occurred, it will show the records that have been modified, clicking on the record will show the containers that have been created and linked.
tc_upload_download_4

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have authority to submit this code.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ktamaral** [comment]
Hello, When you have a chance could you provide a recommendation on how to fix the backend test that has failed?

The backend test passes on our side because the instance type "unspecified" exists in our database. Is there another placeholder instance type we should use instead? Or is there a recommended way to add the "unspecified" instance type to the database such that the test passes?

Here is the backend test:
backend/spec/lib_bulk_import_top_container_linker_validator_spec.rb

Also, the frontend and public tests also have failed but don't appear to be related to any changes we made. Let us know if we need to make any changes on our side for those.

**Comment by lorawoodford** [comment]
@ktamaral For the backend test, please use one of the default/standard `instance_instance_type` values here: https://github.com/archivesspace/archivesspace/blob/master/common/locales/enums/en.yml#L521-L533

@quoideneuf and I will discuss who'll take the lead on the rest of the review, but for the time being that should resolve that backend test issue.

Also, on quick first pass, it looks like all the changes to `public/` were unintentional/unrelated to this work and should probably be removed?

Thanks!

**Comment by rwendler** [comment]
FYI, I am writing some user documentation to accompany this function. I hope to have a draft to share by next week.

**Comment by rwendler** [comment]
@ktamaral How about mixed_materials for the test default?

**Comment by ktamaral** [comment]
Thank you for the help! I've updated the spec test to use 'mixed_materials'. I re-ran the backend test and it still works on our side. Also, I have reverted the files in public/locales/config back to the current version of archivesspace/master. Let us know if there are any questions. Thanks!

File: frontend/app/views/jobs/_show_templates.html.erb [hunk 1]
```diff
@@ -132,6 +132,15 @@
 
 <% end %>
 
+<%# This job type is hidden and will not appear in the job_types list so a template must be created individually %>
+<% define_template "top_container_linker_job", jsonmodel_definition(:top_container_linker_job) do |form, job| %>
+
+    <% begin %>
+      <%= render_aspace_partial :partial => "top_container_linker_job/show", :locals => {:job => job, :form => form} %>
+      <% rescue ActionView::MissingTemplate %>
```
Comment (quoideneuf) [inline frontend/app/views/jobs/_show_templates.html.erb:140]: I am not sure what the intention is here. When I tested the linking job, this `rescue` did happen, since there is no such named template. I recommend taking the entire `begin` clause out if there is indeed no such template.

File: frontend/spec/selenium/spec/locations_spec.rb [hunk 2]
```diff
@@ -93,6 +93,16 @@
     @driver.find_element_with_text('//div[contains(@class, "alert-success")]', /Location Created/)
   end
 
+  it 'allows the browse locations list to be filtered by repository' do
```
Comment (quoideneuf) [inline frontend/spec/selenium/spec/locations_spec.rb:96]: Is this intentional? This test is causing your PR to fail, and appears to be unrelated to the rest of the PR. Perhaps this was included by accident?

**Review by quoideneuf — CHANGES_REQUESTED** [review]
Hi - I made several inline comments and change suggestions, including removing the test that caused the PR to fail the frontend check. I am generally unfamiliar with the existing bulk import code, so my comments are fairly limited. If possible, it would be great if whitespace changes could be eliminated (see for example:https://rnjn.in/2015/09/cleaning-up-whitespace-changes-in-a-pull-request.html). Also, please remove the `.project` line from `.gitignore` as this is specific to a particular IDE and the `.gitignore` is intended for files that are present in all repositories. You can use `.git/info/exclude` or `git config` to hide files that are particular to your development environment. Thanks!

**Comment by quoideneuf** [comment]
(edited comment above to remove the part about `.gitignore` as I see that was changed in one of the commits).

**Comment by ktamaral** [comment]
Hello, I made the updates as recommended. However, for some reason the backend test is now failing (it was passing before). This appears to be failing on a spec file that was not changed in this PR (I confirmed it is the same as the version in archivesspace/master).

rspec ./spec/controller_agent_person_spec.rb:240 # Person agent controller subrecord CRUD gets subrecords along with agent

When you have a chance, let me know what I should do for next steps to get the backend test working again. Perhaps it just needs to be run again? Thank you!

**Comment by ktamaral** [comment]
A quick note that the backend test does pass on my side on the local environment so I am thinking it may need to just be run again. Thanks!

**Comment by ktamaral** [comment]
Apologies for the recent flurry of comments and commits. I pushed a minor commit to try running the backend test again which was failing (but working in our environment), but today it seems that all tests are failing. I also tried reverting back to a previous commit which had all tests passing a few days ago (dae3505), but all tests are failing on that commit now as well. Perhaps there is an issue with the test runner, as indicated by the error messages the build is not failing on all three of them. At this point the branch is up-to-date again including all of the latest requested changes. Note that this is working as expected in our environment so I'm not sure how to troubleshoot this further. Let me know this issue could be something on your side or if there is anything I can do to fix the tests. After the tests are fixed hopefully this should be ready to go. Thank you again!

**Comment by quoideneuf** [comment]
Hi @ktamaral Because of this unusual situation, you (and all other owners of PRs from forked repos) will need to do something to get this unstuck. I think the options are:

1. Add a commit (easiest but not the cleanest)
2. Update the PR's base branch using The API: https://docs.github.com/en/rest/reference/pulls#update-a-pull-request-branch
3. Grant upstream repo permission to modify this PR: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

I prefer #3 if possible since I believe you can set this for all PRs going forward and it would probably be a sensible practice for ArchivesSpace forks.

**Comment by ktamaral** [comment]
Hi @quoideneuf Thank you for your help. In consultation with my team we have decided to go with option #1 and that worked to fix all of the tests! Yay! We did take a look at option #3 although we don't see that option on our side, a few of my colleagues took a look and confirmed (I'm not sure if we're missing something). Let us know there are any questions. We're looking forward to getting this approved. Thanks again!

**Comment by quoideneuf** [comment]
@ktamaral That's fine, it's a small detail. It is there somewhere and if you see it on future PR's please go ahead and tick it so we can rebase when necessary. I'll take a look at this Monday and hopefully get it merged then!

**Review by quoideneuf — APPROVED** [review]
